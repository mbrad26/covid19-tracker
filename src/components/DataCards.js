import React, { useState, useEffect } from 'react';
import { fetchData } from '../api/data.js';

const DataCards = () => {
  const [data, setData] = useState([]);
  const [isLoading, setIsLoading] = useState(false);
  const [isError, setIsError] = useState(false);

  console.log("DATA: ", data);

  useEffect(() => {
    const data = async () => {
      setIsLoading(true);
      try {
        setIsLoading(false);
        const result = await fetchData();
        setData(result.data);
      } catch {
        setIsError(true);
      }
    }
    data();
  }, []);

  return (
    <>
      {isLoading
        ? <h3>Loading ...</h3>
        : data.confirmed ? (
          <>
          <p>Confirmed:<span> {data.confirmed.value.toLocaleString()}</span></p>
          <p>Deaths:<span> {data.deaths.value.toLocaleString()}</span></p>
          <p>Recovered:<span> {data.recovered.value.toLocaleString()}</span></p>
          </>
        )
        : null
      }
      {isError && <h3>Something went wrong ...</h3>}
    </>
  );
};

export default DataCards;
