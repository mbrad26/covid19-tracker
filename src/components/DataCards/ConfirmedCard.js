import React from 'react';
import { Card } from 'react-bootstrap';

import ChartComponent from '../Chart/ChartComponent.js';

const ConfirmedCard = ({ cases, casesIncrease }) => (
  <Card className='shadow' id='cases'>
    <Card.Body>
      <Card.Title>Confirmed: <span className='text-warning font-weight-bold'>
          +{casesIncrease.toLocaleString()}
        </span>
      </Card.Title>
      <ChartComponent label={'Cases'} color={'orange'} />
      <Card.Text>
        {cases.toLocaleString()}
      </Card.Text>
    </Card.Body>
  </Card>
)

export default ConfirmedCard;
