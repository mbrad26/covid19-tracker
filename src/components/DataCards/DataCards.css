.card-body {
  padding: 10px;
  padding-bottom: 10px;
  background-color: #E9E8E6;
}

.card-text {
  margin-top: 5px;
}

.card-title.h5 {
  font-weight: bold;
}

#spinner-cards {
  margin: 100px;
}

i {
  float: right;
  font-size: 2rem;
  margin-right: 5px;
}

#card-cases .card-title i,
#card-cases .card-title span {
  color: #ff7a33;
}

#card-deaths .card-title i,
#card-deaths .card-title span {
  color: #bd0026;
}

#card-recovered .card-title i,
#card-recovered .card-title span {
  color: #92a324;
}

#card-cases,
#card-deaths,
#card-recovered {
  min-width: 275px;
  padding: 5px;
  margin: auto;
}

#update {
  display: inline-flex;
  padding-right: 15px;
  float: right;
  font-variant: small-caps;
}
