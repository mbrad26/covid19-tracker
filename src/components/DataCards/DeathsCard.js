import React from 'react';
import { Card } from 'react-bootstrap';
import CountUp from 'react-countup';

import ChartComponent from '../Chart/ChartComponent.js';

const DeathsCard = ({
  deaths,
  country,
  lastUpdated,
  deathsIncrease,
  historicalData
  }) => {
  const labels = historicalData && Object.keys(historicalData);
  const values = historicalData && Object.values(historicalData);

  return (
    <Card>
      <Card.Body>
        <Card.Title>{country}: <span>
          <CountUp
            prefix='+'
            start={0}
            end={deathsIncrease}
            duration={2.5}
            separator=','
          />
          </span>
          <i className="fas fa-skull-crossbones"></i>
        </Card.Title>
        <ChartComponent
          labels={labels}
          values={values}
          label={'Deaths'}
          color={'#bd0026'}
        />
        <Card.Text>
          {deaths.toLocaleString()}
          <span className='text-muted' id='update'>{lastUpdated}</span>
        </Card.Text>
      </Card.Body>
    </Card>
  );
};

export default DeathsCard;
