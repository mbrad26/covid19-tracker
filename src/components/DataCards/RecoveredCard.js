import React from 'react';
import { Card } from 'react-bootstrap';
import CountUp from 'react-countup';

import ChartComponent from '../Chart/ChartComponent.js';

const RecoveredCard = ({
  country,
  recovered,
  lastUpdated,
  historicalData,
  recoveredIncrease,
  }) => {
  const labels = historicalData && Object.keys(historicalData);
  const values = historicalData && Object.values(historicalData);

  return (
    <Card className='shadow'>
      <Card.Body>
        <Card.Title>{country}: <span className='text-success font-weight-bold'>
          <CountUp
            prefix='+'
            start={0}
            end={recoveredIncrease}
            duration={2.5}
            separator=','
          />
          </span>
        </Card.Title>
        <ChartComponent
          labels={labels}
          values={values}
          label={'Recovered'}
          color={'#228B22'}
        />
        <Card.Text>
          {recovered.toLocaleString()}
          <span id='update'>{lastUpdated}</span>
        </Card.Text>
      </Card.Body>
    </Card>
  );
};

export default RecoveredCard;
