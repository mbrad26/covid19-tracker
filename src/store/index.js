import { createStore, applyMiddleware } from 'redux';
import createSagaMiddleware from 'redux-saga';
import { createLogger } from 'redux-logger';
import rootReducer from '../reducers';
import rootSaga from '../sagas';

const logger = createLogger();
const saga = createSagaMiddleware();

const store = createStore(
  rootReducer,
  applyMiddleware(saga, logger)
);

// console.log('STORE: ', store.getState());

saga.run(rootSaga);

export default store;
