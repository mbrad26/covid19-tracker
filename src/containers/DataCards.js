import { connect } from 'react-redux';

import DataCards from '../components/DataCards/DataCards';
import { doDataLoading } from '../actions/globalData';
import { doHistoricalDataLoading } from '../actions/historicalData';
import { getIsLoadingStatus, getData, getIsErrorStatus } from '../selectors/data';

const mapStateToProps = ({ globalDataState, historicalDataState }) => ({
  globalData: getData(globalDataState),
  isError: getIsErrorStatus(globalDataState),
  historicalData: getData(historicalDataState),
  isLoading: getIsLoadingStatus(globalDataState),
});

const mapDispatchToProps = dispatch => ({
  loadingGlobalData: () => dispatch(doDataLoading()),
  loadingHistoricalData: () => dispatch(doHistoricalDataLoading()),
});

export default connect(mapStateToProps, mapDispatchToProps)(DataCards);
