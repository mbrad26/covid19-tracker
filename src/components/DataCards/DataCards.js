import React, { useEffect } from 'react';
import { Spinner, Row, Col } from 'react-bootstrap';

import './DataCards.css';
import DeathsCard from './DeathsCard';
import ConfirmedCard from './ConfirmedCard';
import RecoveredCard from './RecoveredCard';

const DataCards = ({
  isError,
  isLoading,
  globalData,
  historicalData,
  loadingGlobalData,
  loadingHistoricalData,
  }) => {

  useEffect(() => {
    loadingGlobalData();
    loadingHistoricalData();
  }, [loadingGlobalData, loadingHistoricalData]);

  return (
    <div>
      {isLoading && <Spinner id='spinner-cards' animation="border" variant="light" role="status" />}
      {globalData.cases &&
        <Row id='holder'>
          <Col md={12} lg={4} className='' id='card-cases'>
            <ConfirmedCard
              country={globalData.country ? globalData.country : 'Worldwide'}
              lastUpdated={new Date(globalData.updated).toDateString()}
              cases={globalData.cases}
              casesIncrease={globalData.todayCases}
              historicalData={
                historicalData.timeline
                ? historicalData.timeline.cases
                : historicalData.cases
              }
            />
          </Col>
          <Col md={12} lg={4} className='' id='card-deaths'>
            <DeathsCard
              country={globalData.country ? globalData.country : 'Worldwide'}
              lastUpdated={new Date(globalData.updated).toDateString()}
              deaths={globalData.deaths}
              deathsIncrease={globalData.todayDeaths}
              historicalData={
                historicalData.timeline
                ? historicalData.timeline.deaths
                : historicalData.deaths
              }
            />
          </Col>
          <Col md={12} lg={4} className='' id='card-recovered'>
            <RecoveredCard
              country={globalData.country ? globalData.country : 'Worldwide'}
              lastUpdated={new Date(globalData.updated).toDateString()}
              recovered={globalData.recovered}
              recoveredIncrease={globalData.todayRecovered}
              historicalData={
                historicalData.timeline
                ? historicalData.timeline.recovered
                : historicalData.recovered
              }
            />
            </Col>
        </Row>
      }
      {isError && <h3>Something went wrong ...</h3>}
    </div>
  );
};

export default DataCards;
