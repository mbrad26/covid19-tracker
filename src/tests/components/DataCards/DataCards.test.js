import React from 'react';
import { Provider } from 'react-redux';
import { render, screen, act } from '@testing-library/react';
import axios from 'axios';

import store from '../../../store';
import { globalData, historicalData } from '../../fixtures';
import DataCards from '../../../components/DataCards/DataCards';

jest.mock('axios');

describe('DataCards', () => {
  it('renders snapshot', () => {
    const dataCardsProps = {
      isError: false,
      isLoading: false,
      globalData: globalData,
      historicalData: historicalData,
      loadingGlobalData: jest.fn(),
      loadingHistoricalData: jest.fn(),
    };

    const { container } = render(
      <Provider store={store} >
        <DataCards {...dataCardsProps}/>
      </Provider>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('displays an error message when fetching data fails', () => {
    const dataCardsProps = {
      isError: true,
      isLoading: false,
      globalData: {},
      historicalData: {},
      loadingGlobalData: jest.fn(),
      loadingHistoricalData: jest.fn(),
    };

    render(
      <Provider store={store} >
        <DataCards {...dataCardsProps}/>
      </Provider>
    );

    screen.debug();

    expect(screen.getByText(/Something went wrong/)).toBeInTheDocument();
  });
});
