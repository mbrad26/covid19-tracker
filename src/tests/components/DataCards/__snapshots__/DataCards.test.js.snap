// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataCards renders snapshot 1`] = `
<div>
  <div
    class="row"
    id="holder"
  >
    <div
      class="col-lg-4 col-md-12"
      id="card-cases"
    >
      <div
        class="card"
      >
        <div
          class="card-body"
        >
          <div
            class="card-title h5"
          >
            Worldwide
            : 
            <span>
              <span>
                +0
              </span>
            </span>
            <i
              class="fas fa-frown-open"
            />
          </div>
          <canvas
            class="chartjs-render-monitor"
            height="0"
            style="display: block;"
            width="0"
          />
          <p
            class="card-text"
          >
            25,694,157
            <span
              class="text-muted"
              id="update"
            >
              Tue Sep 01 2020
            </span>
          </p>
        </div>
      </div>
    </div>
    <div
      class="col-lg-4 col-md-12"
      id="card-deaths"
    >
      <div
        class="card"
      >
        <div
          class="card-body"
        >
          <div
            class="card-title h5"
          >
            Worldwide
            : 
            <span>
              <span>
                +0
              </span>
            </span>
            <i
              class="fas fa-skull-crossbones"
            />
          </div>
          <canvas
            class="chartjs-render-monitor"
            height="0"
            style="display: block;"
            width="0"
          />
          <p
            class="card-text"
          >
            855,941
            <span
              class="text-muted"
              id="update"
            >
              Tue Sep 01 2020
            </span>
          </p>
        </div>
      </div>
    </div>
    <div
      class="col-lg-4 col-md-12"
      id="card-recovered"
    >
      <div
        class="card"
      >
        <div
          class="card-body"
        >
          <div
            class="card-title h5"
          >
            Worldwide
            : 
            <span>
              <span>
                +0
              </span>
            </span>
            <i
              class="fas fa-smile"
            />
          </div>
          <canvas
            class="chartjs-render-monitor"
            height="0"
            style="display: block;"
            width="0"
          />
          <p
            class="card-text"
          >
            17,994,215
            <span
              class="text-muted"
              id="update"
            >
              Tue Sep 01 2020
            </span>
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
